proto/p4/p4runtime.proto

```diff
@@ -191,157 +187,50 @@ message Action {
 message ActionProfileGroup {
-  uint32 group_id = 1;
-  // This will likely go away (become part of the P4 program)
+  uint32 action_profile_id = 1;
+  uint32 group_id = 2;
   enum Type {
     UNSPECIFIED = 0;
     SELECT = 1;
     FAST_FAILOVER = 2;
   }
-  Type type = 2;
+  Type type = 3;
   message Member {
     uint32 member_id = 1;
-    int32 weight = 2;  // For WCMP
-    uint32 watch = 3;  // For fast-failover, this is TBD
+    int32 weight = 2;
+    uint32 watch = 3;
   }
-  repeated Member members = 3;
-  // Maximum size of the group (to help provision resources on the target).
-  int32 max_size = 4;
+  repeated Member members = 4;
+  int32 max_size = 5;
 }
 
 //------------------------------------------------------------------------------
-message CounterWriteRequest {
-  uint64 device_id = 1;
-  repeated CounterEntry updates = 2;
-}
-
-message CounterWriteResponse {
-  // Corresponds one-to-one to request.updates, and populated only
-  // if one or more updates didn't succeed.
-  repeated .google.rpc.Status errors = 1;
-}
-
-message CounterReadRequest {
-  uint64 device_id = 1;
-  // Read all counters if empty.
-  // Read all cells of counter.counter_id if counter.cells is empty.
-  // 'CounterEntry' is shared between request and response messages. For
-  // the request, counters.cells.data need not be populated, and is ignored.
-  repeated CounterEntry counters = 2;
-}
-
-message CounterEntry {
-  uint32 counter_id = 1;
-  repeated CounterCell cells = 2;
-  // TODO timestamp?
-}
-
-message CounterCell {
-  oneof type {
-    // For indirect counter
-    int64 index = 1;
-    // For direct counter, the associated table entry match.
-    // table_entry.action is ignored. Other fields specify the match.
-    TableEntry table_entry = 2;
-  }
-  CounterData data = 3;
-}
-
-message CounterData {
-  int64 byte_count = 1;
-  int64 packet_count = 2;
-}
-
-message CounterReadResponse {
-  CounterEntry counter_entry = 1;
-  // Set to true if no responses will follow this one.
-  bool complete = 2;
-}
-
-//------------------------------------------------------------------------------
-message MeterWriteRequest {
-  uint64 device_id = 1;
-  repeated MeterEntry updates = 2;
-}
-
-message MeterWriteResponse {
-  // Corresponds one-to-one to request.updates, and populated only
-  // if one or more updates didn't succeed.
-  repeated .google.rpc.Status errors = 1;
-}
-
-message MeterReadRequest {
-  uint64 device_id = 1;
-  // Read all meters if empty.
-  // Read all cells of meter.meter_id if meter.cells is empty.
-  // 'MeterEntry' is shared between request and response messages. For
-  // the request, meters.cells.config need not be populated, and is ignored.
-  repeated MeterEntry meters = 2;
-}
-
+// For WriteRequest, Update.Type has following semantics:
+//  - INSERT: Insert the cell in the meter array (with given configuration)
+//  - MODIFY: Modify the meter cell configuration
```

Comment:
I still have trouble understanding the semantics of `INSERT` vs `MODIFY` for counters and meters. It seems that in my case at least, the implementation would be exactly the same.
But I guess that applies to tables as well. Would it make sense to have a single idempotent write operation that would replace `INSERT` and `MODIFY`? I think it may make sense from the controller point-of-view but may make switch implementation more difficult...